jgrapht-core/src/main/java/org/jgrapht/util/SubSplitQueue.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * (C) Copyright 2018-2018, by Daniel Mock and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.util;
+
+import java.util.Iterator;
+
+/**
+ * A queue on a bounded set of integers that is splittable. The order of the elements cannot be
+ * changed after instantiation. An element cannot be added after instantiation.
+ * 
```

Comment:
I guess I'm wondering why you can't just use LinkedHashSet<Integer> for this.  All of the necessary operations are constant-time (you can use an iterator for peek/poll), it already has the necessary doubly-linked list logic, and its memory usage will be a lot better for large numbers of vertices.

Reply:
Thank you for your input. We were not aware of the LinkedHashSet! I adapted the code now to work with the LinkedHashSet (see last commits).
However, we are unsatisfied that we have to precompute the ordered adjacency lists for each vertex. (It is necessary that the splitterVertices for the split method are sorted by the same ordering as the LinkedHashList, otherwise the resulting bucket would not have the same order).

Is there another way to do this?

Reply:
Sure. For the LBFS variants LBFS+ and LBFS*, the order inside the buckets has to be preserved. Hence when updateBuckets(v) is invoked (for a vertex v), the neighbors of v have to be processed by this order to guarantee that the newly added buckets are again sorted by this order.

For this the neighborhoods have to be sorted. Since we want to have linear time complexity, it seems necessary to do this at the beginning of the algorithm (see computeSortedNeighborhoods). This results in having a higher space consumption.

This does not affect the implementation of the "normal" LBFS algorithm.